tests/test_dynamodb_full_table_primary_and_hash_key_projections.py

```diff
@@ -0,0 +1,117 @@
+                'tap-dynamodb.expression': "{\"#cmt\": \"Comment\", \"#name\": \"Name\"}", # `expression` field for reserve word.
+                'tap-mongodb.projection': expected_config['ProjectionExpression']
+            }}]
+            connections.select_catalog_and_fields_via_metadata(conn_id,
+                                                               stream_catalog,
+                                                               annotated_schema,
+                                                               additional_md)
+
+        # run full table sync
+        sync_job_name = runner.run_sync_mode(self, conn_id)
+
+        exit_status = menagerie.get_exit_status(conn_id, sync_job_name)
+        menagerie.verify_sync_exit_status(self, exit_status, sync_job_name)
+
+        # verify the persisted schema was correct
+        messages_by_stream = runner.get_records_from_target_output()
+
+        expected_pks = {}
+
+        for config in table_configs:
+            key = {config['HashKey']}
+            if config.get('SortKey'):
+                key |= {config.get('SortKey')}
+            expected_pks[config['TableName']] = key
+
+        # assert that each of the streams that we synced are the ones that we expect to see
+        record_count_by_stream = runner.examine_target_output_file(self,
+                                                                   conn_id,
+                                                                   {x['TableName'] for x in table_configs},
+                                                                   expected_pks)
+
+        state = menagerie.get_state(conn_id)
+
+        first_versions = {}
+
+        # assert that we get the correct number of records for each stream
+        for config in table_configs:
+            table_name = config['TableName']
+
+            self.assertEqual(config['num_rows'],
+                             record_count_by_stream[table_name])
+
+            # assert that an activate_version_message is first and last message sent for each stream
+            self.assertEqual('activate_version',
+                             messages_by_stream[table_name]['messages'][0]['action'])
+            self.assertEqual('activate_version',
+                             messages_by_stream[table_name]['messages'][-1]['action'])
+
+            # assert that the state has an initial_full_table_complete == True
+            self.assertTrue(state['bookmarks'][table_name]['initial_full_table_complete'])
+            # assert that there is a version bookmark in state
+            first_versions[table_name] = state['bookmarks'][table_name]['version']
+            self.assertIsNotNone(first_versions[table_name])
+
+            # assert that the projection causes the correct fields to be returned
+            for message in messages_by_stream[table_name]['messages']:
+                if message['action'] == 'upsert':
+                    if not message['data'].get('_sdc_deleted_at'):
```

Comment:
I don't think skipping the records if the `_sdc_deleted_at` key is present is a good approach here or in `tests/test_dynamodb_log_based_primary_and_hash_key_projections.py`. If somehow this key was added in full table we would be skipping the assertion below which is the most important part of this test. Deleted records are typically only supported in logical replication. In that test we should expect this key to be present and be None. But I don't think we should skip the assertion if that is not the case. We should hard fail if it doesn't match our expectations.

Reply:
@kspeer825 The `sdc_deleted_at` is added by the tap and it would be present only if the record is deleted, else not. For example, when a record is deleted on the aws side and we run the log-based sync, it would retrieve:
`{"type": "RECORD", "stream": "Test", "record": {"Comment": "No-", "SongTitle": "Call-", "_sdc_deleted_at": "2022-01-31T09:28:58.000000Z"}, "version": 1638176328161}`.
However, the `sdc_deleted_at` wont be present in the record at all if it is not deleted.
For example:
`{"type": "RECORD", "stream": "Test", "record": {"Comment": "No-", "test1": ["", "Cookies", "Coffee", 3.14159, {"Age": 35, "Name": "Joe"}], "test.Age": {"z": {"Age": 35, "Name": "Joe"}, "tests": ["Cookies", "Coffee", 3.14159, ["abc", ["xyz"]]], "Age": {"B": 35}, "Name": "Joe1"}, "SongTitle": "Call-"}, "version": 1642684904662}`
Hence, we have kept it as is, as it is present in all the test files because of the above mentioned reason.

Reply:
I think including the `_sdc_deleted_at` column in the schema when using full table replication disagrees with our product requirements. My understanding is that there is no way to track deleted records outside of using logical replication, so if it is included in the schema when full table replication is selected the client will end up with an unnecessary column in their destination full of Null values. 
For the projection test that does use logical replication, I think the implementation leaves a possibility for false negatives. If no `upsert` messages are replicated, or if somehow a deleted record is replicated, the test would not execute the final assertion and it would not fail.

Reply:
@kspeer825 In the schema we do not have the `sdc_deleted_at` field, as it is a dynamic schema. For Example:
`{"type": "SCHEMA", "stream": "Test", "schema": {"type": "object"}, "key_properties": ["Comment", "SongTitle"]}`.
In the following assertion also, it makes the assertion if the `sdc_deleted_at` field is not present in the record, as it is written by the tap in LOG_BASED method (if the record is deleted from the dynamodb database) while writing the record.